rfcs/enforce-legibility.md

```diff
@@ -0,0 +1,80 @@
+## Summary
+[summary]: #summary
+
+This RFC aims to introduce rules for limiting the usage of opens in OCaml code (`open Foo`) and automatically enforce these rules, using a dedicated tool.
+The infrastructure for enforcing these styles rules would also give way to the introduction of other legibility rules in the future if there is a consensus.
+For every such rule, the proposed infrastructure allows for an *incremental* process of compliance upgrade.
+
+## Motivation
+[motivation]: #motivation
+
+Currently, for a newcomer to either the whole code base or one of its components,
+it can be tedious to understand what is going on in the implementation of a feature. This can extend the development time of a new feature or facilitate the introduction of bugs if the code logic is not well understood.
+
+One of the reason is the heavy use of `open`s which impedes the reader's ability to understand where a value is coming from when used. Developer tools such as `merlin` can help locating the origin of a value but it quickly becomes tedious to review the origin of every ambiguous value, wheras directly qualifying the value instead of opening its module would immediatly resolve the problem.
+
+Of course, `open` is a useful tool and there are many cases where it makes sense to use it, such as for importing the infix operators of a module, or for library layering. This RFC aims to define simple, agreed on rules to determine when an `open` is problematic and either *remove* it (and re-qualify every imported value), *move* it closer to its usage, *transform* it into a local open or *restrain it* by explicitly expressing what symbols of the opened module are used in the source file.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Rules
+
+The rules are up to change, but we propose the following very basic rule as a first step, which only defines when an `open` should be completely removed.
+
+**Remove** an `open` if:
+- it is used 5 times or less in its scope
```

Comment:
Thank you for your comments!

> This number seems arbitrary. What proportion of `open`s in the codebase fail this test?

It *is* arbitrary. Currently, excluding all `open`s that match the `keep` rule, 68 opens fail this narrow test.

> Does this limitation solve the problem described in motivation? I understood the problem statement as 'it's hard to know where names are resolving to', which doesn't seem to be made any easier at the 5th use of an open. [...]

I feel like my way of writing this RFC has been a bit clumsy. The rules I present here correspond to what the tool is currently capable of but definitely not the the whole extent of what is planned. Let me explain here and edit the RFC to clarify what I imagine the tool will do. The first matching rule of the following ones would apply:
- delete useless opens (the current rule). This by itself doesn't fully address the motivation
- transform `open`s to local `let open ... in` when their use is local to one or a few functions
- use structured opens like you suggest, when an open is used for less than `X` different symbols
- move `open`s closer to their use. For example, fail the test if an `open` is more than `X` lines away from its optimal placement.

In all cases, these are necessarily arbitrary heuristics, with arbitrary numbers that can be determined by consensus or modified later. In my opinion, this set of rules *would* address the initial motivation, either by reducing the number of opened modules in scope at any point or by re-qualifying values directly.

> It also doesn't look like your tool tracks shadowing (yet?)

Can you explain what kind of shadowing/use case you're talking about? For `open`s that shadow existing identifiers that are later used, I'm currently assuming the warning generated by the compiler is enough.

Reply:
> Let me explain here and edit the RFC to clarify what I imagine the tool will do. The first matching rule of the following ones would apply:

Thanks, this makes more sense to me now.

> * delete useless opens (the current rule). This by itself doesn't fully address the motivation

I was interested to see what the 'useless opens' are, but I haven't been able to compile the tool yet (I need to set up an appropriate switch). I'm still not convinced of the utility of this, but it's hard to say without looking at the code before/after to compare.

> * transform `open`s to local `let open ... in` when their use is local to one or a few functions
> * use structured opens like you suggest, when an open is used for less than `X` different symbols
> * move `open`s closer to their use. For example, fail the test if an `open` is more than `X` lines away from its optimal placement.

I'm a big fan of these :+1:

> Can you explain what kind of shadowing/use case you're talking about? For `open`s that shadow existing identifiers that are later used, I'm currently assuming the warning generated by the compiler is enough.

There are some cases where values are used from one `open`, then another `open` is used and values of the same name are used for that new `open`. I don't remember seeing compiler errors for these, and they have made certain complex parts of the codebase harder to read for me, but I should probably find a specific example before we put effort into this direction.

Reply:
> [...] I haven't been able to compile the tool yet (I need to set up an appropriate switch).

I'm working on fixing some upstream dependencies of ocaml-close so that they are compatible with base 0.13, so soon you should be able to just install it in the mina switch without messing with the already installed packages.


> There are some cases where values are used from one `open`, then another `open` is used and values of the same name are used for that new `open`. I don't remember seeing compiler errors for these, and they have made certain complex parts of the codebase harder to read for me, but I should probably find a specific example before we put effort into this direction.

I was referring to warnings 44 and 45 of the compiler, for which I found surprisingly few documentation, except in the man page : 
```
44   Open statement shadows an already defined identifier.
45   Open statement shadows an already defined label or constructor.
```
And in practice I found that they are not triggered in cases I expected them to. So I agree that it would be interesting to perform that analysis as well.